### ISSUE TITLE
CLI deadlock 

### ISSUE THREAD

**Issue by shaded-enmity** [kennethreitz/pipenv#303]
I'm seeing a deadlock when executing `pipenv install` (full report https://github.com/leapp-to/prototype/pull/15#issuecomment-290447496)

Symptoms:
Executing `pipenv install` never returns.

Cause:
`pipenv` executes `pip` via the `subprocess` module, but doesn't consume the captured output on stdout+stderr:
```bash 
root     20069  0.4  2.2 309120 23044 pts/0    Sl+  14:32   0:00  |                       \_ /root/.local/venvs/pipenv/bin/python3 /root/.local/bin/pipenv install --ignore-hashes
root     20076  0.6  3.2 253584 32636 pts/0    S+   14:32   0:00  |                           \_ /root/.local/share/virtualenvs/le-app-xOFUapXL/bin/python3 /root/.local/share/virtualenvs/le-app-xOFUapXL/bin/pip install -r /tmp/tmpuamy4fy2-requirements.txt -i https://pypi.python.org/simple


[root@centos7-ci vagrant]# strace -p 20069
Process 20069 attached
wait4(20076, 

---

[root@centos7-ci vagrant]# strace -p 20076 -s 1024
Process 20076 attached
write(1, "    Skipping link https://pypi.python.org/packages/bd/bc/d754e01270adc77eb0c645078f90a654030ccd87d7433d62298ec2bf5632/pyasn1-0.1.5-py2.4.egg#md5=9801921c021171460260886117abcd30 (from https://pypi.python.org/simple/pyasn
```

Notice how the `pipenv` process is waiting on the child `pip` process while the child `pip` process is waiting for the parent to consume it's stdout (since the PIPE buffer has filled).

Temporary workaround is patching cli.py and vendor/delegator.py:
https://github.com/leapp-to/prototype/blob/54144592cf7ca6a88eaf8c08946a2ef2aa17864b/ansible/centos-ci/patches/delegator.patch
https://github.com/leapp-to/prototype/blob/54144592cf7ca6a88eaf8c08946a2ef2aa17864b/ansible/centos-ci/patches/cli.patch

I could submit a PR with required changes, but first it'd be good to agree on the overall strategy how to handle this scenario (since the workaround suppresses the output of pip) - I'd probably launch a second thread per each `delegator.run` invocation which would act as stdout/err collector on behalf of the caller.

**Comment by nateprewitt** [comment]
Thanks for this @shaded-enmity! So this seems to be a much more thoroughly fleshed out version of #278. I'm still unable to reproduce a deadlock case locally but the strace seems to show the problem pretty clearly.

Looking into the official documentation for subprocess, this seems to be [a known issue](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait) with using `wait()` on `Popen`. `delegator.py` uses `wait()` by default which seems to be why people are hitting this with a lower threshold for the PIPE buffer filling up. I *think* we may be able to solve this by moving to the suggested `communicate` model suggested by the docs, but we'll need to see if this can be solved completely in delegator.py.

If not, I think your suggested approach seems reasonable, but I'd rather avoid patching pipenv for a problem that seems to be coming from an upstream we control. @kennethreitz, I'm not completely clear on all the tasks you intend `delegator.py` to be capable of handling. Would you have time to look into this? If not, I'll try to slot some time for it asap.

**Comment by shaded-enmity** [comment]
@nateprewitt Digging into this a bit more, my original suggestion with using consumer thread doesn't make much sense since `delegator` runs are sequential, so the fix should be much simpler then that.
I've put together the following minimal reproducer of the issue:
```
import os, subprocess
WRITE_SIZE = 2**16
# this is used in delegator.py
os.environ['PYTHONUNBUFFERED'] = '1'

code = 'import os ; os.write(1, open("/dev/urandom", "rb").read({SIZE}))'.format(SIZE=WRITE_SIZE)
proc = subprocess.Popen(['python3', '-c' , code], stdout=subprocess.PIPE)

print(proc.wait())
print(len(proc.stdout.read()))
```
This should succeed, if you you bump up the `WRITE_SIZE` by 1 byte, it'll most likely fail, but that depends on the underlying kernel pipe buffer size, which most commonly is 16 pages (16*4096=2^16=64KiB).
I was a bit puzzled by `os.environ['PYTHONUNBUFFERED'] = '1'`, so just to verify that this isn't something Python specific, I wrote a C reproducer as well:

```
#define _GNU_SOURCE
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>

#define WRITE_SIZE "64KiB"
#define READ_BUFFER_SZ 1 << 15

int main(int argc, char *argv[]) {
    int rw[2];
    pid_t pid;
    int status;
    ssize_t r;
    char *buffer = malloc(READ_BUFFER_SZ);

    pipe2(rw, O_CLOEXEC);
    pid = fork();

    if (pid == 0) {
        close(rw[0]);
        dup2(rw[1], 1);
        close(rw[1]);
        execl("/usr/bin/head", 
              "/usr/bin/head", "-c", WRITE_SIZE, "/dev/urandom", NULL);
    } else {
        close(rw[1]);
        waitpid(pid, &status, 0);
        while (1) {
            r = read(rw[0], buffer, READ_BUFFER_SZ);
            if (r < 1)
                break;
        }
    }

    free(buffer);

    return 0;
}
```

Which behaves exactly the same way as the Python reproducer.

The fix is rather easy, we just need to change the order of operations from wait -> read to read -> wait, I'll open a PR later today.

**Comment by nateprewitt** [comment]
Great! Thanks for taking the time to really dig into this, @shaded-enmity. It'll be great to finally get this patched in delegator.py.

**Comment by shaded-enmity** [comment]
I think the fix will require discussion since it'll most likely break existing users code.
@nateprewitt  @kennethreitz are you guys available on IRC/Gitter for a quick chat?

**Comment by kennethreitz** [comment]
@shaded-enmity (and @nateprewitt) come to freenode #pipenv.

**Comment by jacebrowning** [comment]
I'm not sure if this is the same issue, but both of these Windows CI builds are timing out now with pipenv 3.5.6:

https://ci.appveyor.com/project/jacebrowning/template-python-demo/build/1.0.539
https://ci.appveyor.com/project/jacebrowning/env-diff/build/1.0.24

**Comment by nateprewitt** [comment]
@jacebrowning it seems it may be related. There's currently a patch (kennethreitz/delegator.py#20) in the works to fix this. Hopefully we'll have something out soon.

**Comment by nateprewitt** [comment]
This is now resolved in pipenv's master branch with eeb22ee. Thanks again @shaded-enmity!
